packages/block-editor/src/components/inserter/hooks/use-block-type-impressions.native.js

```diff
@@ -0,0 +1,55 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect, useDispatch } from '@wordpress/data';
+import { setBlockTypeImpressions } from '@wordpress/react-native-bridge';
+
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../../../store';
+
+function useBlockTypeImpressions( blockTypes ) {
+	const { blockTypeImpressions, enableEditorOnboarding } = useSelect(
+		( select ) => {
+			const { getSettings: getBlockEditorSettings } = select(
+				blockEditorStore
+			);
+			const { editorOnboarding, impressions } = getBlockEditorSettings();
+
+			return {
+				blockTypeImpressions: impressions,
+				enableEditorOnboarding: editorOnboarding,
+			};
+		},
+		[]
+	);
+	const { updateSettings } = useDispatch( blockEditorStore );
+
+	const items = enableEditorOnboarding
+		? blockTypes.map( ( blockType ) => ( {
+				...blockType,
+				isNew: blockTypeImpressions[ blockType.name ] > 0,
+		  } ) )
+		: blockTypes;
+
+	const trackBlockTypeSelected = ( { name } ) => {
+		if ( blockTypeImpressions[ name ] > 0 ) {
+			const updatedBlockTypeImpressions = {
+				...blockTypeImpressions,
+				[ name ]: 0,
+			};
+			// Persist block type impression to JavaScript store
+			updateSettings( {
+				impressions: updatedBlockTypeImpressions,
+			} );
+
+			// Persist block type impression count to native app store
+			setBlockTypeImpressions( updatedBlockTypeImpressions );
```

Comment:
There are a couple of questions I had in giving this PR a last look. **I do not view these issues as blockers for merging this PR.** If they are changes we want to make, I think it would be better to make them in a separate PR.

One question was whether we should have to use both `updateSettings` and `setBlockTypeImpressions` every time we need to update one of the block type impressions values. It feels like we're sidestepping redux here instead of working with it. 

Did we give any thought to finding a way to have an action that changes the impressions settings (`updateSettings({impressions: ...})` kick off a call to `setBlockTypeImpressions` as a side effect? It feels like that approach would be more in keeping with the redux architecture and would avoid any risk of someone deciding to update the block impressions in some other way than using this hook and forgetting to also call `setBlockTypeImpressions` to update the value on the native side. In essence, I think what I'm wondering is whether we should be using a [controller](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#controls-2) to update native anytime a relevant `updateSettings` action is dispatched.

That made me wonder if we should be doing something similar when we "select" the impressions value from the store: i.e., should the selector, when the store does not have any impressions, trigger [a resolver](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#resolvers) that retrieves the impressions from native (and updates the store). In essence, it again kind of feels like by having our components firing off side effects, we're bypassing redux instead of working with it.

If we made those changes, our components would then only interact with the redux store (via actions and selectors) and all of the side effects would be handled within redux by the controllers and resolvers instead of having the side effects handled imperatively alongside the redux selectors and actions in our components. 

I'm a bit out of my depth here (I've barely touched controllers or resolvers before), so I wouldn't be at all surprised if I'm misunderstanding some things. 🙂

Reply:
> One question was whether we should have to use both `updateSettings` and `setBlockTypeImpressions` every time we need to update one of the block type impressions values. 

I agree that managing the JavaScript and Native stores separately anytime a mutation is needed is not ideal. The current implementation was based upon prior implementations, e.g. [retrieving the editor capabilities](https://github.com/WordPress/gutenberg/blob/74a99b2cfce30a49d4c007960f24314ba1a1983e/packages/editor/src/components/provider/index.native.js#L147-L151), and discussions with the team on internal posts. I will share the internal posts separately. 

> In essence, I think what I'm wondering is whether we should be using a [controller](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#controls-2) to update native anytime a relevant `updateSettings` action is dispatched.

Ah, interesting. I was not aware of `controls`. From reviewing the documentation briefly, it does seem appropriate for managing the asynchronous data flow to the Native store. 

> That made me wonder if we should be doing something similar when we "select" the impressions value from the store: i.e., should the selector, when the store does not have any impressions, trigger [a resolver](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/#resolvers) that retrieves the impressions from native (and updates the store).

I was also not aware of `resolvers` either. Given the description of "If your selector result may need to be fulfilled from an external source," it does seem to be a very appropriate way to manage retrieving the data from the Native store. 

> If we made those changes, our components would then only interact with the redux store (via actions and selectors) and all of the side effects would be handled within redux by the controllers and resolvers instead of having the side effects handled imperatively alongside the redux selectors and actions in our components.

Definitely. Having a single interface (i.e. the Redux store selectors and actions) to select and mutate data would be a much better approach. The `controls` and `resolvers` options do look like promising ways of accomplishing that. Thank you for sharing this information. I think it'd be great to refactor this, and possibly other implementations like the Featured Image to leverage these options in the future.